File: python/storage.py
from __future__ import annotations

import json
import sqlite3
import threading
import time
from pathlib import Path
from typing import Any, Callable, TypeVar

T = TypeVar("T")


class Storage:
    def __init__(self, db_path: Path) -> None:
        self._db_path = db_path
        self._lock = threading.Lock()
        self._max_lock_retries = 3
        self._retry_base_delay_s = 0.05

    def initialize(self) -> None:
        def operation(conn: sqlite3.Connection) -> None:
            conn.execute(
                """
                CREATE TABLE IF NOT EXISTS devices (
                    device_id TEXT PRIMARY KEY,
                    platform TEXT NOT NULL,
                    host TEXT NOT NULL,
                    last_seen_at TEXT NOT NULL,
                    capabilities_json TEXT NOT NULL
                )
                """
            )
            conn.execute(
                """
                CREATE TABLE IF NOT EXISTS sessions (
                    session_id TEXT PRIMARY KEY,
                    device_id TEXT NOT NULL,
                    app_id TEXT NOT NULL,
                    platform TEXT NOT NULL,
                    started_at TEXT NOT NULL,
                    ended_at TEXT,
                    status TEXT NOT NULL,
                    trace_id TEXT NOT NULL
                )
                """
            )
            conn.execute(
                """
                CREATE TABLE IF NOT EXISTS perf_samples (
                    id INTEGER PRIMARY KEY AUTOINCREMENT,
                    session_id TEXT NOT NULL,
                    ts TEXT NOT NULL,
                    cpu_pct REAL,
                    memory_mb REAL,
                    launch_ms REAL,
                    fps REAL,
                    jank_pct REAL,
                    metric_flags_json TEXT NOT NULL
                )
                """
            )
            conn.execute(
                """
                CREATE TABLE IF NOT EXISTS artifacts (
                    artifact_id TEXT PRIMARY KEY,
                    session_id TEXT,
                    type TEXT NOT NULL,
                    file_path TEXT NOT NULL,
                    created_at TEXT NOT NULL,
                    meta_json TEXT NOT NULL
                )
                """
            )
            conn.execute(
                """
                CREATE TABLE IF NOT EXISTS audit_logs (
                    id INTEGER PRIMARY KEY AUTOINCREMENT,
                    trace_id TEXT NOT NULL,
                    tool_name TEXT NOT NULL,
                    risk_level TEXT NOT NULL,
                    device_id TEXT,
                    created_at TEXT NOT NULL,
                    result_code TEXT NOT NULL
                )
                """
            )

        self._with_connection(operation)

    def upsert_device(
        self,
        *,
        device_id: str,
        platform: str,
        host: str,
        last_seen_at: str,
        capabilities: dict[str, Any],
    ) -> None:
        capabilities_json = json.dumps(capabilities, separators=(",", ":"))
        def operation(conn: sqlite3.Connection) -> None:
            conn.execute(
                """
                INSERT INTO devices(device_id, platform, host, last_seen_at, capabilities_json)
                VALUES(?, ?, ?, ?, ?)
                ON CONFLICT(device_id)
                DO UPDATE SET
                    platform=excluded.platform,
                    host=excluded.host,
                    last_seen_at=excluded.last_seen_at,
                    capabilities_json=excluded.capabilities_json
                """,
                (device_id, platform, host, last_seen_at, capabilities_json),
            )

        self._with_connection(operation)

    def create_session(
        self,
        *,
        session_id: str,
        device_id: str,
        app_id: str,
        platform: str,
        trace_id: str,
        started_at: str | None = None,
    ) -> None:
        effective_started_at = started_at if started_at is not None else datetime_now_utc_iso()
        def operation(conn: sqlite3.Connection) -> None:
            conn.execute(
                """
                INSERT INTO sessions(session_id, device_id, app_id, platform, started_at, ended_at, status, trace_id)
                VALUES(?, ?, ?, ?, ?, NULL, ?, ?)
                """,
                (session_id, device_id, app_id, platform, effective_started_at, "running", trace_id),
            )

        self._with_connection(operation)

    def close_session(self, *, session_id: str, ended_at: str, status: str) -> None:
        def operation(conn: sqlite3.Connection) -> None:
            conn.execute(
                """
                UPDATE sessions
                SET ended_at=?, status=?
                WHERE session_id=?
                """,
                (ended_at, status, session_id),
            )

        self._with_connection(operation)

    def insert_perf_sample(
        self,
        *,
        session_id: str,
        ts: str,
        cpu_pct: float | None,
        memory_mb: float | None,
        launch_ms: float | None,
        fps: float | None,
        jank_pct: float | None,
        metric_flags: dict[str, str],
    ) -> None:
        metric_flags_json = json.dumps(metric_flags, separators=(",", ":"))
        def operation(conn: sqlite3.Connection) -> None:
            conn.execute(
                """
                INSERT INTO perf_samples(session_id, ts, cpu_pct, memory_mb, launch_ms, fps, jank_pct, metric_flags_json)
                VALUES(?, ?, ?, ?, ?, ?, ?, ?)
                """,
                (session_id, ts, cpu_pct, memory_mb, launch_ms, fps, jank_pct, metric_flags_json),
            )

        self._with_connection(operation)

    def insert_artifact(
        self,
        *,
        artifact_id: str,
        session_id: str | None,
        artifact_type: str,
        file_path: str,
        created_at: str,
        meta: dict[str, Any],
    ) -> None:
        meta_json = json.dumps(meta, separators=(",", ":"))
        def operation(conn: sqlite3.Connection) -> None:
            conn.execute(
                """
                INSERT INTO artifacts(artifact_id, session_id, type, file_path, created_at, meta_json)
                VALUES(?, ?, ?, ?, ?, ?)
                """,
                (artifact_id, session_id, artifact_type, file_path, created_at, meta_json),
            )

        self._with_connection(operation)

    def insert_audit_log(
        self,
        *,
        trace_id: str,
        tool_name: str,
        risk_level: str,
        device_id: str | None,
        created_at: str,
        result_code: str,
    ) -> None:
        def operation(conn: sqlite3.Connection) -> None:
            conn.execute(
                """
                INSERT INTO audit_logs(trace_id, tool_name, risk_level, device_id, created_at, result_code)
                VALUES(?, ?, ?, ?, ?, ?)
                """,
                (trace_id, tool_name, risk_level, device_id, created_at, result_code),
            )

        self._with_connection(operation)

    def list_samples(self, *, session_id: str, limit: int, cursor: int) -> tuple[list[dict[str, Any]], int | None]:
        def operation(conn: sqlite3.Connection) -> list[tuple[Any, ...]]:
            return conn.execute(
                """
                SELECT id, ts, cpu_pct, memory_mb, launch_ms, fps, jank_pct, metric_flags_json
                FROM perf_samples
                WHERE session_id = ? AND id > ?
                ORDER BY id ASC
                LIMIT ?
                """,
                (session_id, cursor, limit),
            ).fetchall()

        rows = self._with_connection(operation)

        samples: list[dict[str, Any]] = []
        next_cursor: int | None = None
        for row in rows:
            sample = {
                "id": row[0],
                "ts": row[1],
                "cpu_pct": row[2],
                "memory_mb": row[3],
                "launch_ms": row[4],
                "fps": row[5],
                "jank_pct": row[6],
                "metric_flags": json.loads(row[7]),
            }
            samples.append(sample)
            next_cursor = row[0]

        return samples, next_cursor

    def list_sessions(self, *, limit: int, cursor: str | None) -> list[dict[str, Any]]:
        query = """
        SELECT session_id, device_id, app_id, platform, started_at, ended_at, status, trace_id
        FROM sessions
        """
        params: list[Any] = []
        if cursor is not None:
            query += " WHERE session_id > ?"
            params.append(cursor)
        query += " ORDER BY session_id ASC LIMIT ?"
        params.append(limit)

        def operation(conn: sqlite3.Connection) -> list[tuple[Any, ...]]:
            return conn.execute(query, params).fetchall()

        rows = self._with_connection(operation)

        return [
            {
                "session_id": row[0],
                "device_id": row[1],
                "app_id": row[2],
                "platform": row[3],
                "started_at": row[4],
                "ended_at": row[5],
                "status": row[6],
                "trace_id": row[7],
            }
            for row in rows
        ]

    def _connect(self) -> sqlite3.Connection:
        self._db_path.parent.mkdir(parents=True, exist_ok=True)
        conn = sqlite3.connect(self._db_path, timeout=5.0)
        conn.execute("PRAGMA journal_mode=WAL")
        conn.execute("PRAGMA busy_timeout=5000")
        return conn

    def _with_connection(self, operation: Callable[[sqlite3.Connection], T]) -> T:
        attempt = 0
        while True:
            try:
                with self._connect() as conn:
                    return operation(conn)
            except sqlite3.OperationalError as exc:
                if not self._is_locked_error(exc) or attempt >= self._max_lock_retries:
                    raise
                time.sleep(self._retry_base_delay_s * (2**attempt))
                attempt += 1

    @staticmethod
    def _is_locked_error(exc: sqlite3.OperationalError) -> bool:
        return "locked" in str(exc).lower()


def datetime_now_utc_iso() -> str:
    from datetime import datetime, timezone

    return datetime.now(timezone.utc).isoformat()

File: src/storage.ts
import { mkdirSync } from "node:fs";
import path from "node:path";
import Database from "better-sqlite3";

const SLEEP_ARRAY = new Int32Array(new SharedArrayBuffer(4));

function sleepMs(durationMs: number): void {
  Atomics.wait(SLEEP_ARRAY, 0, 0, durationMs);
}

function parseJsonRecord(value: string): Record<string, unknown> {
  const parsed = JSON.parse(value);
  if (!parsed || typeof parsed !== "object" || Array.isArray(parsed)) {
    return {};
  }
  return parsed as Record<string, unknown>;
}

export interface CreateSessionInput {
  sessionId: string;
  deviceId: string;
  appId: string;
  platform: string;
  traceId: string;
  startedAt?: string;
}

export interface InsertPerfSampleInput {
  sessionId: string;
  ts: string;
  cpuPct: number | null;
  memoryMb: number | null;
  launchMs: number | null;
  fps: number | null;
  jankPct: number | null;
  metricFlags: Record<string, string>;
}

export interface InsertArtifactInput {
  artifactId: string;
  sessionId: string | null;
  artifactType: string;
  filePath: string;
  createdAt: string;
  meta: Record<string, unknown>;
}

export interface ListSamplesInput {
  sessionId: string;
  limit: number;
  cursor: number;
}

export interface ListSessionsInput {
  limit: number;
  cursor: string | null;
}

export interface PerfSampleRow {
  id: number;
  ts: string;
  cpu_pct: number | null;
  memory_mb: number | null;
  launch_ms: number | null;
  fps: number | null;
  jank_pct: number | null;
  metric_flags: Record<string, unknown>;
}

export interface ListSamplesResult {
  samples: PerfSampleRow[];
  next_cursor: number | null;
}

export interface SessionRow {
  session_id: string;
  device_id: string;
  app_id: string;
  platform: string;
  started_at: string;
  ended_at: string | null;
  status: string;
  trace_id: string;
}

export class Storage {
  private readonly maxLockRetries = 3;
  private readonly retryBaseDelayMs = 50;

  constructor(private readonly dbPath: string) {}

  initialize(): void {
    this.withConnection((db) => {
      db.prepare(
        `
        CREATE TABLE IF NOT EXISTS devices (
            device_id TEXT PRIMARY KEY,
            platform TEXT NOT NULL,
            host TEXT NOT NULL,
            last_seen_at TEXT NOT NULL,
            capabilities_json TEXT NOT NULL
        )
        `
      ).run();

      db.prepare(
        `
        CREATE TABLE IF NOT EXISTS sessions (
            session_id TEXT PRIMARY KEY,
            device_id TEXT NOT NULL,
            app_id TEXT NOT NULL,
            platform TEXT NOT NULL,
            started_at TEXT NOT NULL,
            ended_at TEXT,
            status TEXT NOT NULL,
            trace_id TEXT NOT NULL
        )
        `
      ).run();

      db.prepare(
        `
        CREATE TABLE IF NOT EXISTS perf_samples (
            id INTEGER PRIMARY KEY AUTOINCREMENT,
            session_id TEXT NOT NULL,
            ts TEXT NOT NULL,
            cpu_pct REAL,
            memory_mb REAL,
            launch_ms REAL,
            fps REAL,
            jank_pct REAL,
            metric_flags_json TEXT NOT NULL
        )
        `
      ).run();

      db.prepare(
        `
        CREATE TABLE IF NOT EXISTS artifacts (
            artifact_id TEXT PRIMARY KEY,
            session_id TEXT,
            type TEXT NOT NULL,
            file_path TEXT NOT NULL,
            created_at TEXT NOT NULL,
            meta_json TEXT NOT NULL
        )
        `
      ).run();

      db.prepare(
        `
        CREATE TABLE IF NOT EXISTS audit_logs (
            id INTEGER PRIMARY KEY AUTOINCREMENT,
            trace_id TEXT NOT NULL,
            tool_name TEXT NOT NULL,
            risk_level TEXT NOT NULL,
            device_id TEXT,
            created_at TEXT NOT NULL,
            result_code TEXT NOT NULL
        )
        `
      ).run();
    });
  }

  upsertDevice(input: {
    deviceId: string;
    platform: string;
    host: string;
    lastSeenAt: string;
    capabilities: Record<string, unknown>;
  }): void {
    const capabilitiesJson = JSON.stringify(input.capabilities);
    this.withConnection((db) => {
      db.prepare(
        `
        INSERT INTO devices(device_id, platform, host, last_seen_at, capabilities_json)
        VALUES(?, ?, ?, ?, ?)
        ON CONFLICT(device_id)
        DO UPDATE SET
            platform=excluded.platform,
            host=excluded.host,
            last_seen_at=excluded.last_seen_at,
            capabilities_json=excluded.capabilities_json
        `
      ).run(input.deviceId, input.platform, input.host, input.lastSeenAt, capabilitiesJson);
    });
  }

  createSession(input: CreateSessionInput): void {
    const startedAt = input.startedAt ?? new Date().toISOString();
    this.withConnection((db) => {
      db.prepare(
        `
        INSERT INTO sessions(session_id, device_id, app_id, platform, started_at, ended_at, status, trace_id)
        VALUES(?, ?, ?, ?, ?, NULL, ?, ?)
        `
      ).run(input.sessionId, input.deviceId, input.appId, input.platform, startedAt, "running", input.traceId);
    });
  }

  closeSession(input: { sessionId: string; endedAt: string; status: string }): void {
    this.withConnection((db) => {
      db.prepare(
        `
        UPDATE sessions
        SET ended_at = ?, status = ?
        WHERE session_id = ?
        `
      ).run(input.endedAt, input.status, input.sessionId);
    });
  }

  insertPerfSample(input: InsertPerfSampleInput): void {
    const metricFlagsJson = JSON.stringify(input.metricFlags);
    this.withConnection((db) => {
      db.prepare(
        `
        INSERT INTO perf_samples(session_id, ts, cpu_pct, memory_mb, launch_ms, fps, jank_pct, metric_flags_json)
        VALUES(?, ?, ?, ?, ?, ?, ?, ?)
        `
      ).run(
        input.sessionId,
        input.ts,
        input.cpuPct,
        input.memoryMb,
        input.launchMs,
        input.fps,
        input.jankPct,
        metricFlagsJson
      );
    });
  }

  insertArtifact(input: InsertArtifactInput): void {
    const metaJson = JSON.stringify(input.meta);
    this.withConnection((db) => {
      db.prepare(
        `
        INSERT INTO artifacts(artifact_id, session_id, type, file_path, created_at, meta_json)
        VALUES(?, ?, ?, ?, ?, ?)
        `
      ).run(input.artifactId, input.sessionId, input.artifactType, input.filePath, input.createdAt, metaJson);
    });
  }

  insertAuditLog(input: {
    traceId: string;
    toolName: string;
    riskLevel: string;
    deviceId: string | null;
    createdAt: string;
    resultCode: string;
  }): void {
    this.withConnection((db) => {
      db.prepare(
        `
        INSERT INTO audit_logs(trace_id, tool_name, risk_level, device_id, created_at, result_code)
        VALUES(?, ?, ?, ?, ?, ?)
        `
      ).run(input.traceId, input.toolName, input.riskLevel, input.deviceId, input.createdAt, input.resultCode);
    });
  }

  listSamples(input: ListSamplesInput): ListSamplesResult {
    return this.withConnection((db) => {
      const rows = db
        .prepare(
          `
          SELECT id, ts, cpu_pct, memory_mb, launch_ms, fps, jank_pct, metric_flags_json
          FROM perf_samples
          WHERE session_id = ? AND id > ?
          ORDER BY id ASC
          LIMIT ?
          `
        )
        .all(input.sessionId, input.cursor, input.limit) as Array<{
        id: number;
        ts: string;
        cpu_pct: number | null;
        memory_mb: number | null;
        launch_ms: number | null;
        fps: number | null;
        jank_pct: number | null;
        metric_flags_json: string;
      }>;

      const samples = rows.map((row) => ({
        id: row.id,
        ts: row.ts,
        cpu_pct: row.cpu_pct,
        memory_mb: row.memory_mb,
        launch_ms: row.launch_ms,
        fps: row.fps,
        jank_pct: row.jank_pct,
        metric_flags: parseJsonRecord(row.metric_flags_json)
      }));

      const nextCursor = rows.length > 0 ? rows[rows.length - 1]?.id ?? null : null;
      return {
        samples,
        next_cursor: nextCursor
      };
    });
  }

  listSessions(input: ListSessionsInput): SessionRow[] {
    return this.withConnection((db) => {
      if (input.cursor === null) {
        return db
          .prepare(
            `
            SELECT session_id, device_id, app_id, platform, started_at, ended_at, status, trace_id
            FROM sessions
            ORDER BY session_id ASC
            LIMIT ?
            `
          )
          .all(input.limit) as SessionRow[];
      }

      return db
        .prepare(
          `
          SELECT session_id, device_id, app_id, platform, started_at, ended_at, status, trace_id
          FROM sessions
          WHERE session_id > ?
          ORDER BY session_id ASC
          LIMIT ?
          `
        )
        .all(input.cursor, input.limit) as SessionRow[];
    });
  }

  protected connect(): Database.Database {
    mkdirSync(path.dirname(this.dbPath), { recursive: true });
    const db = new Database(this.dbPath);
    db.pragma("journal_mode = WAL");
    db.pragma("busy_timeout = 5000");
    return db;
  }

  private withConnection<T>(operation: (db: Database.Database) => T): T {
    let attempt = 0;

    while (true) {
      try {
        const db = this.connect();
        try {
          return operation(db);
        } finally {
          db.close();
        }
      } catch (error: unknown) {
        if (!this.isLockedError(error) || attempt >= this.maxLockRetries) {
          throw error;
        }

        sleepMs(this.retryBaseDelayMs * 2 ** attempt);
        attempt += 1;
      }
    }
  }

  private isLockedError(error: unknown): boolean {
    if (!(error instanceof Error)) {
      return false;
    }

    return error.message.toLowerCase().includes("locked");
  }
}
